README.md
Group Tools
===========
Combines different group additions into one plugin

Features
-----------
- group multiple admin
- group activity (profile,dashboard,index,group widget)
	- can show more than one group activity
	- has support for river_comments and like
- group members widget (group only)
- group invitations widget (index, dashboard)
- group mail
- group invite
	- better error messages
	- free text with invite
	- listing of outstanding invitations
	- personal invitation revoke
	- add users to group
	- all users (admin option)
	- import users from csv
- group listing
	- alfabetical
- allow the creation of subgroups

Group default access
-----------------------
As of version 2.3 of this plugin we offer the ability to set a group default access level (like with Site and User), 
this will however not work if you haven't applied https://github.com/Elgg/Elgg/pull/253 to your Elgg installation.

This fix is in place for Elgg 1.9, so no modifications will be needed to the Elgg core files.

CHANGES.txt
= Version History =

2.11 (2014-12-30):

- added: option to create subgroups

x.x:

- added: option for group admins to export the members to CSV (admin enabled option)
- added: bulk group delete admin page
- fixed: PHP fatal exceptions when using group admins on group join
- fixed: problem with some tabs shown when logged out
- fixed: email/domain case sensitivity in invites (thanks to Matt Backett)
- changed: more efficient annotation query for email invitations  (thanks to Matt Backett)

2.10 (2014-05-15):

- added: users can register if registration is disabled but they have a valid group invite code
- fixed: no groups to select in groups_river_widget
- fixed: error during saving of special states
- fixed: if trying to access a hidden group when logged out, correctly forward after login
- fixed: the action URL for user membership killrequest (#57)

2.9.1 (2014-03-14):

- removed: debug code which caused errors to appear on screen

2.9 (2014-03-11):

- added: plugin setting to indicate if a group is hidden
- added: option to configure a group welcome message (#41)
- added: listing of the email invitations for groups (#42)
- added: option to join groups based on email domain (#43)
- added: option to relate groups (#44)
- fixed: layout issue in plugin settings
- fixed: you can no join groups you were invited for on the group profile page
- fixed: misleading translation for a group cleanup setting (#22)
- fixed: create group button showing up on some pages when setting was disabled (#35)

2.8 (2013-10-31):

- added: option to allow group members to invite more members
- added: Portuguese-Brazilian translation (thanks to brunovb)
- added: tags to quickstart discussion widget
- added: suggested groups (based on profile information)
- added: group admins are notified upon membership request #31 (thanks to Matt Beckett)
- fixed: crash when listing a forum topic of a hidden group
- fixed: email invites are not cleaned up on register
- fixed: group owner should not be group admin #26
- changed: better group selection in quickstart discussion widget
- changed: showing group status (open/closed) is a plugin setting #3
- changed: group index widget to include popular groups option #28

2.7 (2013-06-27):

- added: status icon when discussion is closed and fast action to close/open a topic
- added: a widget to quickly start a discussion (index, dashboard and group context)
- fixed: no tab selected in group listing when group_tools is first activated
- fixed: group admin transfer no longer working for non admins
- changed: performance improvement in group admin transfer
- changed: discussion list now shows in which group the discussion takes place (if not already clear)

2.6 (2013-05-06):

- added: option to hide the My Status sidebar
- fixed: using new cache functions
- changed: index groups widget now optionally respects group ordering

2.5 (2013-04-05):

- added: auto join groups based on email invitation when joining the site
- added: different group ACL fixes
- fixed: problem with email invites from hidden groups
- fixed: in some cases email invites weren't saved
- changed: auto notifications now listen to $NOTIFICATION_HANDLERS
- changed: handling of invitations page
- changed: layout of actions
- removed: group creation limited to admins (Elgg core now offers this option)

2.4 (2012-11-06):

- added: filter groups based on a tag field in index_group widget
- fixed: problem with group invite autocomplete on subfolder Elgg installations (thanks to Bruno Grossmann http://community.elgg.org/profile/bruno)
- fixed: unable to edit a group as a normal user when limiting group creation (thanks to buraddo http://community.elgg.org/profile/buraddo)
- changed: group invitations (and add) no always goto mail
- changed: some logic when (re)inviting for better system messages to the users 
- changed: discussion dasboard widget ordering to be inline with default Elgg
- changed: layout of start.php to be more inline with Elgg

2.3 (2012-07-31):

- added: discussion widget from widget manager
- added: group_forum_topics widget from widget manager
- added: group_news widget from widget manager
- added: index_groups widget from widget manager
- added: group sidebar cleanup feature
- added: group profile fallback handling (error on old group profile link & redirect)
- added: group default access (if https://github.com/Elgg/Elgg/pull/253 is applied)
- added: admin action to enable/disable group notifications
- added: option to show the group members in the group related index widgets
- removed: index_discussions widget
- fixed: deprecated notice in invitations overview
- fixed: url error in invitation
- fixed: PHP wanring in group_river_widget
- fixed: pagination in group members sidebar
- fixed: invite autocomplete not using default autocomplete features
- fixed: forum topics widget ordering
- changed: plugin settings layout

2.2 (2012-05-07):

- added: featured group widget
- added: a index widget (for Widget Manager) to show the latest discussions
- added: widget title link for group membership (in combination with Widget Manager)
- added: option to invite all friends or all site members (admin only)
- fixed: outstanding invitations and membership request not always cleaned up
- fixed: incorrect widget title for group members
- fixed: deprecated notices in combination with widget_manager
- fixed: pagination in group members widget
- fixed: wrong action link in remove invitation
- fixed: bug with creating group invitations
- fixed: PHP notices and warnings
- changed: sending of group invite e-mails to be more generic
- changed: restored limited functionality of group activity river

2.1 (2012-04-16):

- added: option to limit group creation to site administrators
- fixed: language error in group sorting (thanks to LeeUK)
- fixed: depricated notice getOwner() => getOwnerGUID()

2.0 BETA 1 (2012-01-17):

- changed: migrated the plugin to Elgg 1.8

1.9 (2011-12-05):


start.php
<?php
/**
 * Start file for this plugin
 */

// define for default group access
define("GROUP_TOOLS_GROUP_ACCESS_DEFAULT", -10);

require_once(dirname(__FILE__) . "/lib/functions.php");
require_once(dirname(__FILE__) . "/lib/events.php");
require_once(dirname(__FILE__) . "/lib/hooks.php");

/**
 * called when the Elgg system get initialized
 *
 * @return void
 */
function group_tools_init() {
	
	// extend css & js
	elgg_extend_view("css/elgg", "css/group_tools/site");
	elgg_extend_view("js/elgg", "js/group_tools/site");
	elgg_extend_view("js/admin", "js/group_tools/admin");
	
	elgg_register_simplecache_view("css/group_tools/autocomplete");
	elgg_register_css("group_tools.autocomplete", elgg_get_simplecache_url("css", "group_tools/autocomplete"));
	
	// extend groups page handler
	elgg_register_plugin_hook_handler("route", "groups", "group_tools_route_groups_handler");
	
	// hook on title menu
	elgg_register_plugin_hook_handler("register", "menu:title", "group_tools_menu_title_handler");
	elgg_register_plugin_hook_handler("register", "menu:user_hover", "group_tools_menu_user_hover_handler");
	elgg_register_plugin_hook_handler("register", "menu:entity", "group_tools_menu_entity_handler");
	
	if (elgg_get_plugin_setting("multiple_admin", "group_tools") == "yes") {
		// add group tool option
		add_group_tool_option("group_multiple_admin_allow", elgg_echo("group_tools:multiple_admin:group_tool_option"), false);
		
		// register permissions check hook
		elgg_register_plugin_hook_handler("permissions_check", "group", "group_tools_multiple_admin_can_edit_hook");
		
		// register on group leave
		elgg_register_event_handler("leave", "group", "group_tools_multiple_admin_group_leave");
		
		//notify admin on membership request
		elgg_register_event_handler("create", "membership_request", "group_tools_membership_request");
	}
		
	// register group activity widget
	// 2012-05-03: restored limited functionality of group activity widget, will be fully restored if Elgg fixes widget settings
	elgg_register_widget_type("group_river_widget", elgg_echo("widgets:group_river_widget:title"), elgg_echo("widgets:group_river_widget:description"), "dashboard,profile,index,groups", true);
	
	// register group members widget
	elgg_register_widget_type("group_members", elgg_echo("widgets:group_members:title"), elgg_echo("widgets:group_members:description"), "groups", false);
	
	// register groups invitations widget
	elgg_register_widget_type("group_invitations", elgg_echo("widgets:group_invitations:title"), elgg_echo("widgets:group_invitations:description"), "index,dashboard", false);
	
	// register featured groups widget
	elgg_register_widget_type("featured_groups", elgg_echo("groups:featured"), elgg_echo("widgets:featured_groups:description"), "index");
	
	// register index groups widget
	elgg_register_widget_type("index_groups", elgg_echo("groups"), elgg_echo("widgets:index_groups:description"), "index", true);
	
	// quick start discussion
	elgg_register_widget_type("start_discussion", elgg_echo("group_tools:widgets:start_discussion:title"), elgg_echo("group_tools:widgets:start_discussion:description"), "index,dashboard,groups");
	
	// group invitation
	elgg_register_action("groups/invite", dirname(__FILE__) . "/actions/groups/invite.php");
	
	// manage auto join for groups
	elgg_extend_view("groups/edit", "group_tools/forms/special_states", 350);
	elgg_register_event_handler("create", "member_of_site", "group_tools_join_site_handler");
	
	// show group edit as tabbed
	elgg_extend_view("groups/edit", "group_tools/group_edit_tabbed", 1);
	elgg_extend_view("groups/edit", "group_tools/group_edit_tabbed_js", 999999999);
	
	// show group profile widgets - edit form
	elgg_extend_view("groups/edit", "group_tools/forms/profile_widgets", 400);
	
	// cleanup group side menu
	elgg_extend_view("groups/edit", "group_tools/forms/cleanup", 450);
	
	// group default access
	elgg_extend_view("groups/edit", "group_tools/forms/default_access");
	
	// group notifications
	elgg_extend_view("groups/edit", "group_tools/forms/notifications", 375);
	
	// allow group members to invite new members
	elgg_extend_view("groups/edit", "group_tools/forms/invite_members", 475);
	
	// configure a group welcome message
	elgg_extend_view("groups/edit", "group_tools/forms/welcome_message");
	
	// configure domain based group join
	elgg_extend_view("groups/edit", "group_tools/forms/domain_based");
	
	// show group status in owner block
	elgg_extend_view("page/elements/owner_block/extend", "group_tools/owner_block");
	// show group status in stats (on group profile)
	elgg_extend_view("groups/profile/summary", "group_tools/group_stats");
	
	if (elgg_is_active_plugin("blog")) {
		elgg_register_widget_type("group_news", elgg_echo("widgets:group_news:title"), elgg_echo("widgets:group_news:description"), "profile,index,dashboard", true);
		elgg_extend_view("css/elgg", "widgets/group_news/css");
		elgg_extend_view("js/elgg", "widgets/group_news/js");
	}
	
	if (elgg_is_admin_logged_in()) {
		run_function_once("group_tools_version_1_3");
	}
	
	// related groups
	add_group_tool_option("related_groups", elgg_echo("groups_tools:related_groups:tool_option"), false);
	elgg_extend_view("groups/tool_latest", "group_tools/modules/related_groups");
	elgg_register_widget_type("group_related", elgg_echo("groups_tools:related_groups:widget:title"), elgg_echo("groups_tools:related_groups:widget:description"), "groups");
	
	// registration
	elgg_extend_view("register/extend", "group_tools/register_extend");
	
	// register index widget to show latest discussions
	elgg_register_widget_type("discussion", elgg_echo("discussion:latest"), elgg_echo("widgets:discussion:description"), "index,dashboard", true);
	elgg_register_widget_type("group_forum_topics", elgg_echo("discussion:group"), elgg_echo("widgets:group_forum_topics:description"), "groups");
	
	// register events
	elgg_register_event_handler("join", "group", "group_tools_join_group_event");
	
	// register plugin hooks
	elgg_register_plugin_hook_handler("widget_url", "widget_manager", "group_tools_widget_url_handler");
	elgg_register_plugin_hook_handler("access:default", "user", "group_tools_access_default_handler");
	elgg_register_plugin_hook_handler("access:collections:write", "user", "group_tools_access_write_handler");
	elgg_register_plugin_hook_handler("action", "groups/join", "group_tools_join_group_action_handler");
	elgg_register_plugin_hook_handler("register", "menu:owner_block", "group_tools_register_owner_block_menu_handler");
	elgg_register_plugin_hook_handler("route", "register", "group_tools_route_register_handler");
	elgg_register_plugin_hook_handler("action", "register", "group_tools_action_register_handler");
	
	// actions
	elgg_register_action("group_tools/admin_transfer", dirname(__FILE__) . "/actions/admin_transfer.php");
	elgg_register_action("group_tools/toggle_admin", dirname(__FILE__) . "/actions/toggle_admin.php");
	elgg_register_action("group_tools/mail", dirname(__FILE__) . "/actions/mail.php");
	elgg_register_action("group_tools/profile_widgets", dirname(__FILE__) . "/actions/profile_widgets.php");
	elgg_register_action("group_tools/cleanup", dirname(__FILE__) . "/actions/cleanup.php");
	elgg_register_action("group_tools/default_access", dirname(__FILE__) . "/actions/default_access.php");
	elgg_register_action("group_tools/invite_members", dirname(__FILE__) . "/actions/invite_members.php");
	elgg_register_action("group_tools/welcome_message", dirname(__FILE__) . "/actions/welcome_message.php");
	elgg_register_action("group_tools/domain_based", dirname(__FILE__) . "/actions/domain_based.php");
	elgg_register_action("group_tools/related_groups", dirname(__FILE__) . "/actions/related_groups.php");
	elgg_register_action("group_tools/remove_related_groups", dirname(__FILE__) . "/actions/remove_related_groups.php");
	elgg_register_action("group_tools/member_export", dirname(__FILE__) . "/actions/member_export.php");
	
	elgg_register_action("group_tools/toggle_special_state", dirname(__FILE__) . "/actions/admin/toggle_special_state.php", "admin");
	elgg_register_action("group_tools/fix_auto_join", dirname(__FILE__) . "/actions/admin/fix_auto_join.php", "admin");
	elgg_register_action("group_tools/notifications", dirname(__FILE__) . "/actions/admin/notifications.php", "admin");
	elgg_register_action("group_tools/fix_acl", dirname(__FILE__) . "/actions/admin/fix_acl.php", "admin");
	elgg_register_action("group_tools/admin/bulk_delete", dirname(__FILE__) . "/actions/admin/bulk_delete.php", "admin");
	
	elgg_register_action("groups/email_invitation", dirname(__FILE__) . "/actions/groups/email_invitation.php");
	elgg_register_action("groups/decline_email_invitation", dirname(__FILE__) . "/actions/groups/decline_email_invitation.php");
	elgg_register_action("group_tools/revoke_email_invitation", dirname(__FILE__) . "/actions/groups/revoke_email_invitation.php");

	elgg_register_action("group_tools/order_groups", dirname(__FILE__) . "/actions/order_groups.php", "admin");
	
	elgg_register_action("discussion/toggle_status", dirname(__FILE__) . "/actions/discussion/toggle_status.php");

	// subpermissions
	elgg_register_action("group_tools/subpermissions/add", dirname(__FILE__) . "/actions/subpermissions/add.php");
	elgg_register_action("group_tools/subpermissions/delete", dirname(__FILE__) . "/actions/subpermissions/delete.php");
	elgg_register_action("group_tools/subpermissions/add_member", dirname(__FILE__) . "/actions/subpermissions/add_member.php");
	elgg_register_action("group_tools/subpermissions/delete_member", dirname(__FILE__) . "/actions/subpermissions/delete_member.php");

	add_group_tool_option("subpermissions", elgg_echo("group_tools:subpermissions:activate"), false);
	elgg_register_event_handler("leave", "group", "group_tools_subpermissions_leave");
	elgg_register_event_handler("delete", "group", "group_tools_subpermissions_delete");
}

/**
 * called when the system is ready
 *
 * @return void
 */
function group_tools_ready() {
	// unregister dashboard widget group_activity
	elgg_unregister_widget_type("group_activity");
}

/**
 * called just before a page starts with output
 *
 * @return void
 */
function group_tools_pagesetup() {
	
	$user = elgg_get_logged_in_user_entity();
	$page_owner = elgg_get_page_owner_entity();
	
	if (elgg_in_context("groups") && ($page_owner instanceof ElggGroup)) {
		if ($page_owner->forum_enable == "no") {
			// unset if not enabled for this plugin
			elgg_unregister_widget_type("group_forum_topics");
		}
		
		if (!empty($user)) {
			// check for admin transfer
			$admin_transfer = elgg_get_plugin_setting("admin_transfer", "group_tools");
			
			if (($admin_transfer == "admin") && $user->isAdmin()) {
				elgg_extend_view("groups/edit", "group_tools/forms/admin_transfer", 400);
			} elseif (($admin_transfer == "owner") && (($page_owner->getOwnerGUID() == $user->getGUID()) || $user->isAdmin())) {
				elgg_extend_view("groups/edit", "group_tools/forms/admin_transfer", 400);
			}
			
			// check multiple admin
			if (elgg_get_plugin_setting("multiple_admin", "group_tools") == "yes") {
				// extend group members sidebar list
				elgg_extend_view("groups/sidebar/members", "group_tools/group_admins", 400);
				
				// remove group tool options for group admins
				if (($page_owner->getOwnerGUID() != $user->getGUID()) && !$user->isAdmin()) {
					remove_group_tool_option("group_multiple_admin_allow");
				}
			}
			
			// invitation management
			if ($page_owner->canEdit()) {
				$request_options = array(
					"type" => "user",
					"relationship" => "membership_request",
					"relationship_guid" => $page_owner->getGUID(),
					"inverse_relationship" => true,
					"count" => true
				);
				
				$requests = elgg_get_entities_from_relationship($request_options);
				
				$postfix = "";
				if (!empty($requests)) {
					$postfix = " [" . $requests . "]";
				}
				
				if (!$page_owner->isPublicMembership()) {
					elgg_register_menu_item("page", array(
						"name" => "membership_requests",
						"text" => elgg_echo("groups:membershiprequests") . $postfix,
						"href" => "groups/requests/" . $page_owner->getGUID(),
					));
				} else {
					elgg_register_menu_item("page", array(
						"name" => "membership_requests",
						"text" => elgg_echo("group_tools:menu:invitations") . $postfix,
						"href" => "groups/requests/" . $page_owner->getGUID(),
					));
				}
			}

			// subpermissions management
			if ($page_owner->canEdit() && $page_owner->subpermissions_enable == "yes") {
				elgg_register_menu_item("page", array(
					"name" => "subpermissions",
					"text" => elgg_echo("group_tools:subpermissions"),
					"href" => "groups/subpermissions/" . $page_owner->getGUID(),
					));
			}
			
			// group mail options
			if ($page_owner->canEdit() && (elgg_get_plugin_setting("mail", "group_tools") == "yes")) {
				elgg_register_menu_item("page", array(
					"name" => "mail",
					"text" => elgg_echo("group_tools:menu:mail"),
					"href" => "groups/mail/" . $page_owner->getGUID(),
				));
			}
		}
	}
	
	if ($page_owner instanceof ElggGroup) {
		if (!$page_owner->isPublicMembership()) {
			if (elgg_get_plugin_setting("search_index", "group_tools") != "yes") {
				// closed groups should be indexed by search engines
				elgg_extend_view("page/elements/head", "metatags/noindex");
			}
		}
		
		// cleanup sidebar
		elgg_extend_view("page/elements/sidebar", "group_tools/sidebar/cleanup");
	}
	
	
	elgg_register_admin_menu_item("administer", "group_bulk_delete", "administer_utilities");
}

/**
 * Upgrade script to fix some problem
 *
 * @return void
 */
function group_tools_version_1_3() {
	$dbprefix = elgg_get_config("dbprefix");
	
	$query = "SELECT ac.id AS acl_id, ac.owner_guid AS group_guid, er.guid_one AS user_guid
		FROM {$dbprefix}access_collections ac
		JOIN {$dbprefix}entities e ON e.guid = ac.owner_guid
		JOIN {$dbprefix}entity_relationships er ON ac.owner_guid = er.guid_two
		WHERE e.type = 'group'
		AND er.relationship = 'member'
		AND er.guid_one NOT IN (
			SELECT acm.user_guid
			FROM {$dbprefix}access_collections ac2
			JOIN {$dbprefix}access_collection_membership acm ON ac2.id = acm.access_collection_id
			WHERE ac2.owner_guid = ac.owner_guid
		)";
	
	$data = get_data($query);
	if (!empty($data)) {
		foreach ($data as $row) {
			add_user_to_access_collection($row->user_guid, $row->acl_id);
		}
	}
}

// default elgg event handlers
elgg_register_event_handler("init", "system", "group_tools_init");
elgg_register_event_handler("ready", "system", "group_tools_ready");
elgg_register_event_handler("pagesetup", "system", "group_tools_pagesetup", 550);
